pkgpanda/util.py

```diff
@@ -264,13 +265,32 @@ def make_file(name):
 
 
 def write_json(filename, data):
-    with open(filename, "w+") as f:
-        return json.dump(data, f, **json_prettyprint_args)
+    dumped_json = json.dumps(data, **json_prettyprint_args)
+    write_string(filename, dumped_json)
 
 
 def write_string(filename, data):
-    with open(filename, "w+") as f:
-        return f.write(data)
+    """
+    Write a string to a file.
+    Overwrite any data in that file.
+
+    We use an atomic write practice of creating a temporary file and then
+    moving that temporary file to the given ``filename``. This prevents race
+    conditions such as the file being read by another process after it is
+    opened here but not yet written to.
+
+    It also prevents us from creating or truncating a file before we fail to
+    write data to it because of low disk space.
+    """
+    temporary_filename = '{filename}-{random}.tmp'.format(
```

Comment:
Can we please use just the basename of the file in here?

I am afraid some functions may try to call this functions with a full path instead of just a filename.

Reply:
The important thing is that the temporary file is in the same directory as the destination file. This ensures that it is on the same disk partition, and hence that `os.replace` will be atomic.

Reply:
@vespian That was relevant when I was using `mkstemp` but not now, as per @jongiddy 's comment.